readMe.md
# express-pg-session-next
A customizable PostgreSQL session store for Connect / Express.
**This is a fixed version** of [express-pg-session](https://github.com/sanjaypojo/express-pg-session).
The original library is [express-pg-session](https://www.npmjs.com/package/express-pg-session).

## Forked from connect-pg-simple
Most of the code is borrowed from [connect-pg-simple](https://github.com/voxpelli/node-connect-pg-simple)

## Main change
The only real change is the ability to customize the column names in your session tables. This can be done by supplying a columns option to pgSession. This is a frequent requirement in our workflow and so we made this change.

```javascript
let columnNames = {
  session_id: 'sid',
  session_data: 'sess',
  expire: 'expires_at'
}

let session = new pgSession({
  pool : pgPool,                // Connection pool
  tableName : 'user_sessions',  // Alternate table name
  columns: columnNames          // Alternate column names
})
```

## Installation

```bash
npm install express-pg-session
```

Once npm installed the module, you need to create the **session** table in your database. For that you can use the [table.sql](table.sql) file provided with the module:

```bash
psql mydatabase < node_modules/connect-pg-simple/table.sql
```

Or simply play the file via a GUI, like the pgAdminIII queries tool.

## Usage

Examples are based on Express 4.

Simple example:

```javascript
var session = require('express-session');

app.use(session({
  store: new (require('express-pg-session')(session))(),
  secret: process.env.FOO_COOKIE_SECRET,
  resave: false,
  cookie: { maxAge: 30 * 24 * 60 * 60 * 1000 } // 30 days
}));
```

Advanced example showing some custom options:

```javascript
var pg = require('pg')
  , session = require('express-session')
  , pgSession = require('express-pg-session')(session);

var pgPool = new pg.Pool({
    // Insert pool options here
});

app.use(session({
  store: new pgSession({
    pool : pgPool,                // Connection pool
    tableName : 'user_sessions'   // Use another table-name than the default "session" one
  }),
  secret: process.env.FOO_COOKIE_SECRET,
  resave: false,
  cookie: { maxAge: 30 * 24 * 60 * 60 * 1000 } // 30 days
}));
```

Express 3 (and similar for Connect):

```javascript
var express = require('express');

app.use(session({
  store: new (require('express-pg-session')(express.session))(),
  secret: process.env.FOO_COOKIE_SECRET,
  cookie: { maxAge: 30 * 24 * 60 * 60 * 1000 } // 30 days
}));
```

## Advanced options

* **pool** - Recommended. Connection pool object (compatible with [pg.Pool](https://github.com/brianc/node-pg-pool)) for the underlying database module. The **conString** option is ignored if this option is specified.
* **pgPromise** - Existing instance of `pg-promise` to be used for DB communications. The **conString** option is ignored if this option is specified.
* **conString** - If you don't specify a pool object, use this option or `conObject` to specify a PostgreSQL connection [string](https://github.com/brianc/node-postgres/wiki/Client#new-clientstring-url-client) and this module will create a new pool for you. If the connection string is in the `DATABASE_URL` environment variable (as you do by default on eg. Heroku) – then this module fallback to that if this option is not specified.
* **conObject** - If you don't specify a pool object, use this option or `conString` to specify a PostgreSQL Pool connection [object](https://github.com/brianc/node-postgres#pooling-example) and this module will create a new pool for you.
* **ttl** - the time to live for the session in the database – specified in seconds. Defaults to the cookie maxAge if the cookie has a maxAge defined and otherwise defaults to one day.
* **schemaName** - if your session table is in another Postgres schema than the default (it normally isn't), then you can specify that here.
* **tableName** - if your session table is named something else than `session`, then you can specify that here.
* **pruneSessionInterval** - sets the delay in seconds at which expired sessions are pruned from the database. Default is `60` seconds. If set to `false` no automatic pruning will happen. Automatic pruning weill happen `pruneSessionInterval` seconds after the last pruning – manual or automatic.
* **errorLog** – the method used to log errors in those cases where an error can't be returned to a callback. Defaults to `console.error()`, but can be useful to override if one eg. uses [Bunyan](https://github.com/trentm/node-bunyan) for logging.

## Useful methods

* **close()** – if this module used its own database module to connect to Postgres, then this will shut that connection down to allow a graceful shutdown.
* **pruneSessions([callback(err)])** – will prune old sessions. Only really needed to be called if **pruneSessionInterval** has been set to `false` – which can be useful if one wants improved control of the pruning.


CHANGELOG.md
# Changelog 📝

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.1.3] - 2022-04-13

### Changed

* Security: updates old and vulnerable dependencies

## [1.1.2] - 2022-04-13

### Added

* `CHANGELOG.md`

## [1.1.1] - 2022-04-12

### Changed

* Fixed missing `$` in [touch method](https://github.com/diego-betto/express-pg-session-next/blob/master/src/index.js#L279)


[1.1.3]: https://github.com/diego-betto/express-pg-session-next/releases/tag/v1.1.3
[1.1.2]: https://github.com/diego-betto/express-pg-session-next/releases/tag/v1.1.2
[1.1.1]: https://github.com/diego-betto/express-pg-session-next/releases/tag/v1.1.1

src/index.js
'use strict';

const util = require('util');
const oneDay = 86400;

const currentTimestamp = function () {
  return Math.ceil(Date.now() / 1000);
};

module.exports = function (session) {
  const Store = session.Store || session.session.Store;

  const PGStore = function (options) {
    options = options || {};
    Store.call(this, options);

    this.schemaName = options.schemaName || null;
    this.tableName = options.tableName || 'session';
    this.columns = options.columns || {
      expire: 'expire',
      session_id: 'sid',
      session_data: 'sess'
    };

    this.ttl = options.ttl;

    this.errorLog = options.errorLog || console.error.bind(console);

    if (options.pool !== undefined) {
      this.pool = options.pool;
      this.ownsPg = false;
    } else if (options.pgPromise !== undefined) {
      if (typeof options.pgPromise.query !== 'function') {
        throw new Error('`pgPromise` config must point to an existing and configured instance of pg-promise pointing at your database');
      }
      this.pgPromise = options.pgPromise;
      this.ownsPg = false;
    } else {
      const conString = options.conString || process.env.DATABASE_URL;
      let conObject = options.conObject;

      if (!conObject) {
        conObject = {};

        if (conString) {
          conObject.connectionString = conString;
        }
      }

      this.pool = new (require('pg')).Pool(conObject);
      this.pool.on('error', err => {
        this.errorLog('PG Pool error:', err.message);
      });
      this.ownsPg = true;
    }

    if (options.pruneSessionInterval === false) {
      this.pruneSessionInterval = false;
    } else {
      this.pruneSessionInterval = (options.pruneSessionInterval || 60) * 1000;
      setImmediate(function () {
        this.pruneSessions();
      }.bind(this));
    }
  };

  /**
   * Inherit from `Store`.
   */

  util.inherits(PGStore, Store);

  /**
   * Closes the session store
   *
   * Currently only stops the automatic pruning, if any, from continuing
   *
   * @access public
   */

  PGStore.prototype.close = function () {
    this.closed = true;

    if (this.pruneTimer) {
      clearTimeout(this.pruneTimer);
      this.pruneTimer = undefined;
    }

    if (this.ownsPg) {
      this.pool.end();
    }
  };

  /**
   * Does garbage collection for expired session in the database
   *
   * @param {Function} [fn] - standard Node.js callback called on completion
   * @access public
   */

  PGStore.prototype.pruneSessions = function (fn) {
    this.query(`DELETE FROM ${this.quotedTable()} WHERE ${this.columns.expire} < to_timestamp($1)`, [currentTimestamp()], function (err) {
      if (fn && typeof fn === 'function') {
        return fn(err);
      }

      if (err) {
        this.errorLog('Failed to prune sessions:', err.message);
      }

      if (this.pruneSessionInterval && !this.closed) {
        if (this.pruneTimer) {
          clearTimeout(this.pruneTimer);
        }
        this.pruneTimer = setTimeout(this.pruneSessions.bind(this, true), this.pruneSessionInterval);
        this.pruneTimer.unref();
      }
    }.bind(this));
  };

  /**
   * Get the quoted table.
   *
   * @return {String} the quoted schema + table for use in queries
   * @access private
   */

  PGStore.prototype.quotedTable = function () {
    let result = '"' + this.tableName + '"';

    if (this.schemaName) {
      result = '"' + this.schemaName + '".' + result;
    }

    return result;
  };

  /**
   * Figure out when a session should expire
   *
   * @param {Number} [maxAge] - the maximum age of the session cookie
   * @return {Number} the unix timestamp, in seconds
   * @access private
   */

  PGStore.prototype.getExpireTime = function (maxAge) {
    let ttl = this.ttl;

    ttl = ttl || (typeof maxAge === 'number' ? maxAge / 1000 : oneDay);
    ttl = Math.ceil(ttl + currentTimestamp());

    return ttl;
  };

  /**
   * Query the database.
   *
   * @param {String} query - the database query to perform
   * @param {(Array|Function)} [params] - the parameters of the query or the callback function
   * @param {Function} [fn] - standard Node.js callback returning the resulting rows
   * @access private
   */

  PGStore.prototype.query = function (query, params, fn) {
    if (!fn && typeof params === 'function') {
      fn = params;
      params = [];
    }

    if (this.pgPromise) {
      this.pgPromise.query(query, params || [])
        .then(function (res) { fn && fn(null, res && res[0] ? res[0] : false); })
        .catch(function (err) { fn && fn(err, false); });
    } else {
      this.pool.query(query, params || [], function (err, res) {
        if (fn) { fn(err, res && res.rows[0] ? res.rows[0] : false); }
      });
    }
  };

  /**
   * Attempt to fetch session by the given `sid`.
   *
   * @param {String} sid – the session id
   * @param {Function} fn – a standard Node.js callback returning the parsed session object
   * @access public
   */

  PGStore.prototype.get = function (sessionId, fn) {
    this.query(
      `SELECT ${this.columns.session_data} FROM ${this.quotedTable()}
      WHERE ${this.columns.session_id} = $1 AND expire >= to_timestamp($2)`,
      [sessionId, currentTimestamp()],
      function (err, data) {
        if (err) { return fn(err); }
        if (!data) { return fn(); }
        try {
          return fn(null, (typeof data[this.columns.session_data] === 'string') ? JSON.parse(data[this.columns.session_data]) : data[this.columns.session_data]);
        } catch (e) {
          return this.destroy(sessionId, fn);
        }
      }.bind(this));
  };

  /**
   * Commit the given `sess` object associated with the given `sid`.
   *
   * @param {String} sid – the session id
   * @param {Object} sess – the session object to store
   * @param {Function} fn – a standard Node.js callback returning the parsed session object
   * @access public
   */

  PGStore.prototype.set = function (sessionId, sessionData, fn) {
    const expireTime = this.getExpireTime(sessionData.cookie.maxAge);
    const query1 = `
      UPDATE ${this.quotedTable()}
      SET ${this.columns.session_data} = $1, expire = to_timestamp($2)
      WHERE ${this.columns.session_id} = $3
      RETURNING ${this.columns.session_id}
    `;

    this.query(query1, [sessionData, expireTime, sessionId], (err, data) => {
      if (!err && data === false) {
        const query2 = `
          INSERT INTO ${this.quotedTable()}
          (
            ${this.columns.session_data},
            ${this.columns.expire},
            ${this.columns.session_id}
          )
          SELECT $1, to_timestamp($2), $3
          WHERE NOT EXISTS (
            SELECT 1 FROM ${this.quotedTable()}
            WHERE ${this.columns.session_id} = $4
          )
        `;

        this.query(query2, [sessionData, expireTime, sessionId, sessionId], function (err) {
          if (fn) { fn.apply(this, err); }
        });
      } else {
        if (fn) { fn.apply(this, err); }
      }
    });
  };

  /**
   * Destroy the session associated with the given `sid`.
   *
   * @param {String} sid – the session id
   * @access public
   */

  PGStore.prototype.destroy = function (sessionId, fn) {
    this.query(
      `DELETE FROM ${this.quotedTable()} WHERE ${this.columns.session_id} = $1`,
      [sessionId],
      function (err) {
        if (fn) { fn(err); }
      }
    );
  };

  /**
   * Touch the given session object associated with the given session ID.
   *
   * @param {String} sid – the session id
   * @param {Object} sess – the session object to store
   * @param {Function} fn – a standard Node.js callback returning the parsed session object
   * @access public
   * 
   * @description fixes "$" on rows 282 and 283
   */

  PGStore.prototype.touch = function (sessionId, sessionData, fn) {
    const expireTime = this.getExpireTime(sessionData.cookie.maxAge);

    this.query(
      `UPDATE ${this.quotedTable()}
        SET ${this.columns.expire} = to_timestamp($1)
        WHERE ${this.columns.session_id} = $2
        RETURNING ${this.columns.session_id}
      `,
      [expireTime, sessionId],
      function (err) { fn(err); }
    );
  };

  return PGStore;
};
